### PR TITLE
ENH: add getPVAliases and grep_more_ioc scripts

### PR DESCRIPTION
**Pull request by aberges-SLAC**
## Description
Added a script for reconstructing PV aliases from their child IOC's, `st.cmd` files, and `.archive` files for use in ArchiverAppliance scripts. Does a pythonic `grep_ioc` for the regex pattern supplied, then searches the child IOC for the parent's release pointer. After confirmation, reconstructs all the record ←→ aliases using `.../db/*.archive` and `st.cmd` from the child IOC.

Also added a script called `grep_more_ioc` which is an a personal utility inspired by `grep_ioc` . Most of the legwork of this script is reused in `getPVAlias` so I decided to bundle it for posterity. 

## Motivation and Context
Motors, namely SmarActs and Elliptecs, tend to move around to different areas often and/or have dynamic configurations for channels based on the area's application. This results in PV aliases becoming stale references to channels that no longer house that stage. To rectify this, you'll need to get the currently built association of EPICS records ←→ aliases for the PVs. 

Additionally, I added some QOL in `grep_more_ioc` that helps day-to-day controls work in CLI without having to open the `iocmanager` gui. 

## How Has This Been Tested?
I've run this script & wrapper in an ipython session and using the supplied bash wrapper. Please see the screenshots below!

## Where Has This Been Documented?
All documentation is supplied inline on the scripts with accompanying usage/help doctext.

### getPVAliases
Initial confirmation of ioc caught by a `grep_ioc` like search:
image

All axes/motors generated by the child IOC's `st.cmd` file:
image

Preview of what will be saved to the `.txt` file dumps:
image

Assumes current directory as the base for new folder generation & dump, but accepts different directories:
image

Makes the folder if it does not exist, then iterates to the next IOC:
image

Individual axes can be skipped, with lazy skipping through the rest of the axes:
image

--> Otherwise if you don't accept _all_ axes dumped, (or lazy accept/skip the rest), you will be asked to confirm each dump:
image

Example of output:
image

---------------------------------------------------------------------------------------------------------------------------------
### grep_more_ioc

This script uses `pandas` to print DataFrames, which is quite slower than `prettytable` but doesn't suffer the wrapping/width issues that the latter does. (see below)
image

There are 2 major functions: `print` and `search`. In my examples below I'll use the alias `gmi` for `grep_more_ioc`

#### print
Options:
- `skip_comments`
image
Functions like `cat`, except you skip commented code. Useful for quickly grabbing IOC information when building `happi` entries.

- `release`
image
Searches the child IOCs for the release pointer, then adds an abbreviated path to the parent's current release.

- `print_dirs`
image
Dumps directories to terminal for quicker copy/pasting/`pushd`

#### search
Options:
- `quiet`
image
image
Suppress warnings for paths that don't exist when doing batch searches (as disabled IOCs are often previously deleted...)

- `only`
image
Only print the search results, don't print the DataFrame.

### CODE REVIEW

File: scripts/getPVAliases.sh
Comment (tangkong) [inline scripts/getPVAliases.sh:30]: Were you running into issues sourcing `pcds_conda` after it's been already activated?  I've never had an issue doing so

File: scripts/getPVAliases.sh
Comment (tangkong) [inline scripts/getPVAliases.sh:23]: I believe you can just let python's argparse handle bad args and the help request, if you pass all the arguments through to your python script.

**Review by tangkong — COMMENTED** [review]
Nice work, seems like this should be very helpful!  I've left a couple nitpicky notes, I'll take a bit to look at this from a higher level after this.  

* I like to avoid regex when possible.  It's powerful but often difficult to maintain and parse (by humans).  There's a few places here where I don't think you need it.  
* Thanks for the comments / docstrings!  Love to see it.

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:24]: This is _extremely_ nice, and really which I looked for it before I reinvented the wheel, lol!

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:159]: I very much wanted to avoid `iocmanager`, but was stubborn about the `ioc.cfg` outputs which were just so very nearly `json`.  Doing it this way let me use regex liberally, although it should absolutely have restraint haha.

File: scripts/getPVAliases.sh
Comment (aberges-SLAC) [reply scripts/getPVAliases.sh:30]: It really came down to saving 2-3 seconds per execution in a different script that I based a lot of my ioc manipulation on. That script called `pandas` which is a notoriously long import so I was trying to shave time off as much as I could.

File: scripts/getPVAliases.sh
Comment (aberges-SLAC) [reply scripts/getPVAliases.sh:23]: `source pcds_conda` is greedy and eats `-h` :(

So I tossed this in there as a hacky work-around

File: scripts/getPVAliases.sh
Comment (ZLLentz) [inline scripts/getPVAliases.sh:43]: Feel free to also try something like this if sourcing pcds_conda is ever slow. It's likely that your script doesn't need the full environment sourced, and rather just needs the executable.

This also means you'd be able to do all the arg handling in the python script (if you removed source pcds_conda)
```suggestion
/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.2/bin/python "$ENG_TOOLS_SCRIPTS/getPVAliases.py" $@
```

File: scripts/getPVAliases.py
Comment (ZLLentz) [inline scripts/getPVAliases.py:1]: I'd recommend running all python scripts through `flake8` or `ruff` to check for python styling standards. If you're into having style automatically standardized for you there's also `black`.

I'm pointing this out here because I saw various minor flake8 violations as I was reading through the code.

File: scripts/getPVAliases.sh
Comment (ZLLentz) [inline scripts/getPVAliases.sh:1]: I'd recommend running all shell scripts through shellcheck to check for potential escaping issues etc.

File: scripts/getPVAliases.py
Comment (ZLLentz) [inline scripts/getPVAliases.py:182]: These sorts of lists may get hard to manage over time.

I suggest either adding the missing `txi` or pulling this from the subdirectories present in the /cds/group/pcds/pyps/config directory.

File: scripts/getPVAliases.py
Comment (ZLLentz) [inline scripts/getPVAliases.py:87]: I'm noticing that you've type hinted all the arguments but none of the return values, and none of the lists have their containing types, and a lot of these are type hinted incorrectly (type str set to None here, for example). Are you interested in filling in the missing type information or were you just going for a light pass to help write the script?

File: scripts/getPVAliases.sh
Comment (ZLLentz) [inline scripts/getPVAliases.sh:45]: Is it customary to include a bare exit at the end of a script? What function does it serve? I'm worried this masks the return code from the python call.

File: scripts/getPVAliases.py
Comment (ZLLentz) [reply scripts/getPVAliases.py:159]: On the flip side, this code becomes much easier to maintain if we replicate the native parsing. I could imagine how all of the replace/strip/regex done here may run into some unknown edge case later down the line.

I personally would advocate for changing iocmanager's format to a real json structure but that feels out of scope here.

**Review by ZLLentz — COMMENTED** [review]
Script is good, put in my suggestions and nitpicks

File: scripts/getPVAliases.py
Comment (tangkong) [reply scripts/getPVAliases.py:1]: Should we add a pre-commit config to `engineering_tools`? 🤔

Comment (aberges-SLAC) [reply scripts/getPVAliases.py:1]: Ah, quite a few offenders here in this script. Let me clean that up. I also left in the `#%%` from Spyder for cells that probably don't serve purpose for folks.

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:182]: Good catch! I think that's the safer option. I can just import that as a constant earlier in the script and refer to it and make it better to maintain.

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:87]: I can clean this up. I think I originally left some of these as default = `None` for catching missing args or forcing break points in earlier testing and forgot to revert. I should go around and fill in all the `list[str]` and `list[dict]` etc.

I keep `output : list[str] = None` for reconstructing the `grep_ioc` output that I then parse again, so maybe it's superfluous and I should be more clever about this.

File: scripts/getPVAliases.sh
Comment (aberges-SLAC) [reply scripts/getPVAliases.sh:45]: This is probably just a bad habit from my end. No need to give it a bare `exit`, or any `exit` for that matter

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:159]: I'm a huge fan of that as well, but I agree it's outside of the scope for this script sadly.

I know regex is a spice and it's easy to over season, but I quite like the option of regex searching the innards of the `ioc.cfg` dicts as it can make it much easier for catching bulk info in the MODs, e.g. `lm2k2.*qmini`. Maybe this is also my laziness speaking and it's not a good reason to keep it in, but wouldn't the `execfile` route require you to know/provide the full ioc name explicitly? 

This is probably safer, tbh, since you can't accidentally grab info you don't want. But on the other hand, naming conventions for MCS2 controllers seldom include MODs/TILE info that is relevant (also, sometimes TILEs are stretch across 2 IOCs):
image

So for fixing all the HRM aliases you'd probably want to catch both these IOCs.

What's the smallest hammer I can use here for the same endgoal?

File: scripts/getPVAliases.sh
Comment (aberges-SLAC) [reply scripts/getPVAliases.sh:43]: I very much prefer this, and quite honestly, I'd rather do that and reduce the whole `getPVAliases.sh` to this, since all other arg handling is done by argparser and is now very easy to maintain.

File: scripts/getPVAliases.py
Comment (tangkong) [reply scripts/getPVAliases.py:159]: I think regex is probably a good way to tackle the `ioc.cfg` files, I think we're specifically proposing parsing the iocmanager config file with `execfile`.  Running `execfile` on the individual `ioc-*.cfg` doesn't itself make much sense.

File: scripts/getPVAliases.sh
Comment (aberges-SLAC) [reply scripts/getPVAliases.sh]: Fixed with suggested changes

**Review by tangkong — COMMENTED** [review]
A quick once through, not paying too much attention to the logic (which I had assumed did not change much from my last review).  But after reading through I think I want to look at the details more closely

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:25]: Yeah, I'm not entirely sure _how_ to get around that. I guess I could check in the walk if any valid `.cfg` files exist?

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:43]: It _used_ to, lol, my bad! I forgot to update the doctext on this function when I merged other versions together. Let me fix that in my next commit

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:76]: 🙇

File: scripts/getPVAliases.py
Comment (tangkong) [reply scripts/getPVAliases.py:25]: Checking the contents of the folder was my first instinct, but it's only marginally more stable.

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:95]: I'm actually very partial to pandas for printing to terminal, but the import is a _slog_ and I wanted something a bit faster for this script. Now that I'm importing from `grep_more_ioc` anyway, kind of a moot point and I might as well just use pandas? The only slight annoyance is fixing column with to max alias size but that's trivial.

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:230]: Interesting! I've never tried string concat, frankly, so it's nice to know that it works. Adding to next commit

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:2]: That is mostly an artifact from my Spyder settings and honestly can be removed. The header docstring should probably be changed in both cases. Especially `grep_more_ioc` which used a different, almost arbitrary semantic versioning for my own notes/reference when it was a personal script.

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:32]: Sure! That works — I also could've just imported it from `grep_more_ioc` in the first place as it's used in the `find_ioc` function which depends on this constant.

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:328]: I'm one of those weirdos who prefers it, but it really only has utility when you're at 3+ methods (which, tbf, I should be avoiding in the first place). Added to next commit!

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:421]: I think the sparse nature of argparse (:badumtiss:) including only the invoked parser & subparser made it weird for testing, and I got distracted by other things and didn't leave a note to clean up.

I should probably [do something with subparser.set_default](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.set_defaults) to give me some namespace clues. Something like `subparser.set_default(command='print')` etc.

File: scripts/grep_more_ioc.py
Comment (tangkong) [reply scripts/grep_more_ioc.py:2]: Also the versions of these scripts will be captured in engineering tools' versions anyway

File: scripts/getPVAliases.py
Comment (tangkong) [reply scripts/getPVAliases.py:95]: Not worth changing since it's here already, and you're doing some messing with the formatting.  I was just wondering.

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:25]: Something to the effect of:
```
VALID_HUTCH = ['all']
for (path, dir, files) in os.walk('/cds/group/pcds/pyps/config'):
    if 'iocmanager.cfg' in files:
    VALID_HUTCH.append(path.rsplit(r'/', maxsplit=1)[-1])
sorted(VALID_HUTCH)
```

Comment (tangkong) [reply scripts/getPVAliases.py:25]: I was thinking something silly like
```python
import glob
sorted([d for d in glob.glob(path + '/*/') if glob.glob(d + 'iocmanager.cfg')])
```

File: scripts/getPVAliases.sh
Comment (ZLLentz) [inline scripts/getPVAliases.sh:4]: Two things here:

- Do you know where `$ENG_TOOLS_SCRIPTS` gets set? Mine doesn't even match the engineering tools on my path, so for me this script actually will run a different .py file than the one I intend!
- Maybe it'd be more portable to have this point to the same-directory .py file instead of using the custom environment variables

```suggestion
THIS_DIR="$(dirname "$(realpath "${BASH_SOURCE[0]}")")"
/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.2/bin/python "$THIS_DIR/getPVAliases.py" "$@"
```

**Comment by ZLLentz** [comment]
I'm starting to feel like this has been stuck in review cycle for longer than it needs to be relative to how complete the code is, so I'm going to try to focus my reviews on bigger things related to deploying/sharing/using the scripts. I can't wait to see this in engineering tools proper.

File: scripts/grep_more_ioc.py
Comment (ZLLentz) [reply scripts/grep_more_ioc.py:328]: the `.something` on each new line is really nice when each step returns a thing that you can also call a method on, which happens more in some languages than in others. I don't see this pattern often in Python but it pays off in these situations:
```
my_thing
    .do_thing()
    .do_another_thing()
    .do_more_many_thing()
```

File: scripts/getPVAliases.py
Comment (ZLLentz) [inline scripts/getPVAliases.py:1]: One additional big-picture thing: I recommend adding your new scripts to the readme.

**Comment by aberges-SLAC** [comment]
I _just_ caught a bug that's worth fixing before merging as well. Happens when you `print -c` or `print -r` on the _very_ rare case where a `child_ioc.cfg` doesn't actually exist, but the `ioc` is absolutely valid and in `HUTCH/iocmanager.cfg` *cough PLCs cough*.

I'm really just commenting this here to remind myself to use `try/except` with `search_file` for handling these edge cases.
image

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:25]: Not silly at all! I like this much more. Very clean, much glob

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:421]: Turns out my original approach is one of the few stable ways to check for other possible `args`. When `argparse` parses the args, if any subparser is unused, none of its attributes get populated in the resultant `Namespace` object. 
image

I guess the other way of doing this would be generating a full `Namespace` object with all of the defaults set and to check the `bool` of any `sub-commands` because I already structured them as simple toggles at the top level. 
Something like:

```
class Args:
    self.patt = None
    self.hutch = None
    self.ignore_disabled = False
    ...
````

But then I'm hard coding all the defaults anyway that `argparse` was originally handling. 🤷

**Comment by aberges-SLAC** [comment]
Okay, I fixed the bug when searching IOCs that had `',\n history: [...'` causing the IOC to get skipped during the search. Added a sub-function for printing these `history` attributes within `grep_more_ioc` but I'll add that to the `README.md` in post, since the pre-commit hates the `README.md` syntax.

**Comment by ZLLentz** [comment]
> the pre-commit hates the README.md syntax

What's the story here? Maybe we need to make some of the pre-commit entries ignore the readme?

**Comment by aberges-SLAC** [comment]
> > the pre-commit hates the README.md syntax
> 
> What's the story here? Maybe we need to make some of the pre-commit entries ignore the readme?

It complained loudly at me for markdown syntax that I never edited, and then seemingly wanted to remove lines like `<tr/>` just to add them back. It was somewhere like 6-7 commits ago when I was struggling to get the pre-commit to accept me (turns out adding it to my local repo helps [:).

Just let it its thing and committed the pre-commit changes a few tries ago. Now that it is happy I'm hesitant to make any further modifications. Besides, its much cleaner on the web interface anyhow.

**Comment by tangkong** [comment]
The README.md probably hadn't been pre-commit-checked, since our checks only run on modified files.

File: README.md
Comment (ZLLentz) [inline README.md:251]: The script is actually `grep_more_ioc.sh`, I think we should document as such here or rename the scripts to not have .sh suffixes.

Comment (silkenelson) [reply README.md:251]: I'd vote to not have .sh as most scripts in engineering tools do not have extensions.

Comment (tangkong) [reply README.md:251]: To hop on to the readme suggestions, could you add an example call for those of us who don't know what pattern/hutch to search for?   I think examples are helpful for people not so familiar with EPICS and our systems

**Comment by ZLLentz** [comment]
I ran the scripts myself and asked the questions that I had as a user, maybe they inform documentation/examples/etc.

File: README.md
Comment (ZLLentz) [reply README.md:254]:
```
$ grep_more_ioc.sh --help
usage: grep_more_ioc [-h] [-d] patt hutch {print,search} ...

Transforms grep_ioc output to json object and prints in pandas.DataFrame

positional arguments:
  patt
  hutch
  {print,search}        Follow-up commands after grep_ioc executes
    print               Just a simple print of the dataframe to the terminal.
    search              For using regex-like searches in the child IOC.cfg captured by grep_ioc. Useful for quickly gathering instance information, IP addr, etc.
```

File: scripts/getPVAliases.py
Comment (aberges-SLAC) [reply scripts/getPVAliases.py:260]: I should update this docstring to "Do you want to add all resulting PV <--> alias associations to the final output?"

The script now saves a single file with all the results appended together (less annoying to parse when you're playing with the result later). Since you're iterating through a base PV and its resulting PVs (each with aliases), instead of pestering the user to approve every single instance (e.g. each axis in an MCS2 ioc), the user can approve if the first association looks good.

File: README.md
Comment (aberges-SLAC) [reply README.md:254]: I can improve this in the `argparse` documentation so it is more transparent.

The _annoying_ thing about how `argparse` handles subparsers is that each `subparser` has its own help text, so something like:
`grep_more_ioc.sh . all print --help` triggers the help text for `print`
`grep_more_ioc.sh . all search --help` triggers the help text for `search`

Personally, I hate it, but I haven't found a good enough work around _yet_

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:1]: Yes, they are positional arguments. I contemplated making `grep_more_ioc.sh <patt> <hutch>` default to print but it gets hairy with the subparsers.

File: README.md
Comment (tangkong) [reply README.md:254]: One simple thing you could do is add some text in the top-level parser that hints the user to try those sub-parser help messages.

File: scripts/getPVAliases.py
Comment (ZLLentz) [reply scripts/getPVAliases.py:260]: I think if the script saves a file we should mention this in the docstring.

Comment (aberges-SLAC) [reply scripts/getPVAliases.py:260]: More informative?
image

Comment (ZLLentz) [reply scripts/getPVAliases.py:260]: It's definitely better but I think my overall confusion was a lack of understanding that this script was supposed to save anything at all.

Comment (aberges-SLAC) [reply scripts/getPVAliases.py:260]: With some massaging on `argparse.RawTextHelpFormatter`:
image

File: scripts/grep_more_ioc.py
Comment (aberges-SLAC) [reply scripts/grep_more_ioc.py:1]: Also improved:
image

File: scripts/getPVAliases.py
Comment (ZLLentz) [reply scripts/getPVAliases.py:260]: Looks good to me! Very helpful output!

File: scripts/grep_more_ioc.py
Comment (ZLLentz) [reply scripts/grep_more_ioc.py:1]: This one looks good too!

File: README.md
Comment (ZLLentz) [reply README.md:251]: I think if this script renaming/examples stuff get in then this PR has my approval (I don't want to hold it up any longer, but I'd like the names to be finalized + I'd like people to be ready to use it when it gets merged)

**Comment by aberges-SLAC** [comment]
Okay, all help doc-text improved, committed, and pushed!

**Review by tangkong — DISMISSED** [review]
This looks good to me.  I think future improvements probably deserve their own separate PRs.  This is some good work! 👍

**Review by ZLLentz — CHANGES_REQUESTED** [review]
The final tiny thing before merge is the script names, the majority of the executables in engineering_tools don't have the .sh extension, so I suggest renaming these to remove it.

It's worth noting that you've already documented this without the .sh extension in the readme, so renaming things makes everything consistent.

Otherwise, if there's a good reason to keep the .sh, it should be documented as such in the readme.

**Comment by aberges-SLAC** [comment]
> The final tiny thing before merge is the script names, the majority of the executables in engineering_tools don't have the .sh extension, so I suggest renaming these to remove it.
> 
> It's worth noting that you've already documented this without the .sh extension in the readme, so renaming things makes everything consistent.
> 
> Otherwise, if there's a good reason to keep the .sh, it should be documented as such in the readme.

Done and done! see new commits

**Review by ZLLentz — APPROVED** [review]
Let's get this in!

**Comment by ZLLentz** [comment]
I'm going to merge this now. Post-merge if there are other concerns we can handle them in smaller, reviewable follow-up PRs. We can either tag/deploy this immediately or we can wait, let me know what you all prefer.

**Comment by aberges-SLAC** [comment]
> I'm going to merge this now. Post-merge if there are other concerns we can handle them in smaller, reviewable follow-up PRs. We can either tag/deploy this immediately or we can wait, let me know what you all prefer.

I'm excited to get this out for folks to use & stress test. I'm sure there's more bugs hiding somewhere that's outside of my normal use cases. So I'm in favor of deployment sooner rather than later — the script is pretty stable for now.

Are there any confluences pages I should update as well? Or do we leave the documentation in the README.md?

**Comment by ZLLentz** [comment]
I'll release another version right away then- tags are cheap. Do you want to announce the release of R3.8.0 (which, compared to last version, only has this PR) or would you prefer I do it?

I don't think there's any confluence page in particular that needs to be updated, unless you think a bigger page about usage etc. would be useful.

**Pull request merged**
